aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java

```diff
@@ -0,0 +1,222 @@
+import static io.opentelemetry.semconv.resource.attributes.ResourceAttributes.SERVICE_NAME;
+import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.DB_OPERATION;
+import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.DB_SYSTEM;
+import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.FAAS_INVOKED_NAME;
+import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.FAAS_INVOKED_PROVIDER;
+import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.GRAPHQL_OPERATION_TYPE;
+import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.MESSAGING_OPERATION;
+import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.MESSAGING_SYSTEM;
+import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.PEER_SERVICE;
+import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.RPC_METHOD;
+import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.RPC_SERVICE;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * AwsMetricAttributeGenerator generates very specific metric attributes based on low-cardinality
+ * span and resource attributes. If such attributes are not present, we fallback to default values.
+ *
+ * <p>The goal of these particular metric attributes is to get metrics for incoming and outgoing
+ * traffic for a service. Namely, {@link SpanKind#SERVER} and {@link SpanKind#CONSUMER} spans
+ * represent "incoming" traffic, {@link SpanKind#CLIENT} and {@link SpanKind#PRODUCER} spans
+ * represent "outgoing" traffic, and {@link SpanKind#INTERNAL} spans are ignored.
+ */
+final class AwsMetricAttributeGenerator implements MetricAttributeGenerator {
+
+  private static final Logger logger =
+      Logger.getLogger(AwsMetricAttributeGenerator.class.getName());
+
+  // Special SERVICE attribute value if GRAPHQL_OPERATION_TYPE attribute key is present.
+  private static final String GRAPHQL = "graphql";
+
+  // Default attribute values if no valid span attribute value is identified
+  private static final String UNKNOWN_SERVICE = "UnknownService";
+  private static final String UNKNOWN_OPERATION = "UnknownOperation";
+  private static final String UNKNOWN_REMOTE_SERVICE = "UnknownRemoteService";
+  private static final String UNKNOWN_REMOTE_OPERATION = "UnknownRemoteOperation";
+
+  @Override
+  public Attributes generateMetricAttributesFromSpan(SpanData span, Resource resource) {
```

Comment:
The `Resource` looks like a dependency of this class. Probably better to supply it via a parametrized constructor?

Reply:
Maybe - I actually prefer feeding it into the exporter/processor so it's more clear where it is coming from. I don't think this is a blocking comment, so going to leave this as it is.